synapse/lib/cell.py

```diff
@@ -4032,3 +4048,241 @@ async def _runReloadFunc(self, name, func):
+
+        now = s_common.now()
+        vals = {
+            'shadow': shadow,
+            'updated': now
+        }
+        kdef['shadow'] = shadow
+
+        if duration is None:
+            vals['expires'] = None
+        else:
+            if duration < 1:
+                raise s_exc.BadArg(mesg='Duration must be equal or greater than 1', name='duration')
+            vals['expires'] = now + duration
+
+        kdef.update(vals)
+        kdef = s_schemas.reqValidUserApiKeyDef(kdef)
+
+        await self._push('user:apikey:update', kdef.get('user'), iden, vals)
+
+        logger.info(f'Regenerated HTTP API Key {iden} for {user.name}',
+                    extra=await self.getLogExtra(target_user=user.iden, target_username=user.name, iden=iden,
+                                                 status='MODIFY'))
+
+        kdef.pop('shadow')
+        return token, kdef
+
+    @s_nexus.Pusher.onPush('user:apikey:update')
+    async def _setUserApiKey(self, user, iden, vals):
+        lkey = s_common.uhex(user) + b'apikey' + s_common.uhex(iden)
+        buf = self.slab.get(lkey, db=self.usermetadb)
+        if buf is None:
+            raise s_exc.NoSuchIden(mesg=f'User API Key does not exist: {iden}')
+        kdef = s_msgpack.un(buf)
+        kdef.update(vals)
+        self.slab.put(lkey, s_msgpack.en(kdef), db=self.usermetadb)
+        # TODO Timecache clear
+        return kdef
+
+    async def delUserApiKey(self, iden):
+        kdef = await self.getUserApiKey(iden)
+        if kdef is None:
+            raise s_exc.NoSuchIden(mesg=f'User API Key does not exist: {iden}')
+        useriden = kdef.get('user')
+        user = await self.auth.reqUser(useriden)
+        ret = await self._push('user:apikey:del', useriden, iden)
+        logger.info(f'Deleted HTTP API Key {iden} for {user.name}',
+                    extra=await self.getLogExtra(target_user=user.iden, target_username=user.name, iden=iden,
+                                                 status='MODIFY'))
+        return ret
+
+    @s_nexus.Pusher.onPush('user:apikey:del')
+    async def _delUserApiKey(self, user, iden):
+        user = s_common.uhex(user)
+        iden = s_common.uhex(iden)
+
+        self.slab.pop(iden, db=self.apikeydb)
+        self.slab.pop(user + b'apikey' + iden, db=self.usermetadb)
+        # TODO Timecache clear
+        return True
+
+    async def _onUserDelEvnt(self, evnt):
+        # Call callback for handling user:del events
+        user = evnt[1].get('iden')
+        ukey = s_common.uhex(user)
+        for lkey, buf in self.slab.scanByPref(ukey, db=self.usermetadb):
+            suffix = lkey[16:]
+            # Special handling for certain meta which has secondary indexes
+            if suffix.startswith(b'apikey'):
+                key_iden = suffix[6:]
+                self.slab.pop(key_iden, db=self.apikeydb)
+                # TODO Timecache clear
+            self.slab.pop(lkey, db=self.usermetadb)
```

Comment:
`self.slab.delete` is just a hair quicker if we're not assigning the return value